src/tensors/levels/mutex_levels.jl

```diff
@@ -0,0 +1,272 @@
+    atomics = freshen(ctx, tag, :_locks)
+    push_preamble!(ctx, quote
+            $sym = $ex
+            $atomics = $ex.locks
+        end)
+    lvl_2 = virtualize(ctx, :($sym.lvl), Lvl, sym)
+    temp = VirtualMutexLevel(lvl_2, sym, atomics, typeof(level_fill_value(Lvl)), Val, AVal, Lvl)
+    temp
+end
+
+Base.summary(lvl::VirtualMutexLevel) = "Mutex($(lvl.Lvl))"
+virtual_level_resize!(ctx, lvl::VirtualMutexLevel, dims...) = (lvl.lvl = virtual_level_resize!(ctx, lvl.lvl, dims...); lvl)
+virtual_level_size(ctx, lvl::VirtualMutexLevel) = virtual_level_size(ctx, lvl.lvl)
+virtual_level_ndims(ctx, lvl::VirtualMutexLevel) = length(virtual_level_size(ctx, lvl.lvl))
+virtual_level_eltype(lvl::VirtualMutexLevel) = virtual_level_eltype(lvl.lvl)
+virtual_level_fill_value(lvl::VirtualMutexLevel) = virtual_level_fill_value(lvl.lvl)
+
+function declare_level!(ctx, lvl::VirtualMutexLevel, pos, init)
+    lvl.lvl = declare_level!(ctx, lvl.lvl, pos, init)
+    return lvl
+end
+
+function assemble_level!(ctx, lvl::VirtualMutexLevel, pos_start, pos_stop)
+    pos_start = cache!(ctx, :pos_start, simplify(ctx, pos_start))
+    pos_stop = cache!(ctx, :pos_stop, simplify(ctx, pos_stop))
+    idx = freshen(ctx, :idx)
+    lockVal = freshen(ctx, :lock)
+    push_preamble!(ctx, quote
+              Finch.resize_if_smaller!($(lvl.locks), $(ctx(pos_stop)))
+              @inbounds for $idx = $(ctx(pos_start)):$(ctx(pos_stop))
+                $(lvl.locks)[$idx] = Finch.make_lock(eltype($(lvl.AVal)))
+              end
+          end)
+    assemble_level!(ctx, lvl.lvl, pos_start, pos_stop)
+end
+
+supports_reassembly(lvl::VirtualMutexLevel) = supports_reassembly(lvl.lvl)
+function reassemble_level!(ctx, lvl::VirtualMutexLevel, pos_start, pos_stop)
+    pos_start = cache!(ctx, :pos_start, simplify(ctx, pos_start))
+    pos_stop = cache!(ctx, :pos_stop, simplify(ctx, pos_stop))
+    idx = freshen(ctx, :idx)
+    lockVal = freshen(ctx, :lock)
+    push_preamble!(ctx, quote
+              Finch.resize_if_smaller!($lvl.locks, $(ctx(pos_stop)))
+              @inbounds for $idx = $(ctx(pos_start)):$(ctx(pos_stop))
+                $lvl.locks[$idx] = Finch.make_lock(eltype($(lvl.AVal)))
+              end
+          end)
+    reassemble_level!(ctx, lvl.lvl, pos_start, pos_stop)
+    lvl
+end
+
+function freeze_level!(ctx, lvl::VirtualMutexLevel, pos)
+    idx = freshen(ctx, :idx)
+    push_preamble!(ctx, quote
+        resize!($(lvl.locks), $(ctx(pos)))
+    end)
+    lvl.lvl = freeze_level!(ctx, lvl.lvl, pos)
+    return lvl
+end
+
+function thaw_level!(ctx::AbstractCompiler, lvl::VirtualMutexLevel, pos)
+    lvl.lvl = thaw_level!(ctx, lvl.lvl, pos)
+    return lvl
+end
+
+function virtual_moveto_level(ctx::AbstractCompiler, lvl::VirtualMutexLevel, arch)
+    #Add for seperation level too.
+    atomics = freshen(ctx, :locksArray)
+
+    push_preamble!(ctx, quote
+        $atomics = $(lvl.locks)
+        $(lvl.locks) = $moveto($(lvl.locks), $(ctx(arch)))
```

Comment:
I know I probably wrote something like this in the initial implementation but something about this does strike me as kind of cursed given the intended use of moveto.

Reply:
Ah, yes. I recently fixed #608, and I imagine at some point we will need to get a little more precise about how we use moveto (perhaps explicitly making copies of state variables and other things)